### PR TITLE
[authpolicy-v2] AuthPolicy v1beta2

### LINKED ISSUE #247: Upgrade AuthPolicy inner types to AuthConfig v1beta2
- [x] #253 (keep `AuthRules`)
 - [x] #248 (remove `AuthRule`)

### PR DESCRIPTION
**Pull request by guicassolato**
Base branch for the first chunk of #207.

Closes:
- [x] #262
- [x] #253 → replaces version `v1beta1` of the `AuthPolicy` CRD with new `v1beta2` (based on Authorino's `AuthConfig/v1beta2`)

As well as:
- Via #266
  - [x] #263
  - [x] #264
- Via #256
  - [x] #248
- Via #270
  - [x] #265
- Via #275 
  - [x] AuthPolicy v2 docs and user guides
- Via #279
  - [x] #278 
- Via #281
  - [x] Validation rules (e.g. `targetRef.group/kind`, `routeSelectors` in gw policies)

### How to review/verify this PR

- Learn the basis of it
  - we bumped the version of Authorino and Authorino Operator imported types
  - we migrated the core of the AuthPolicy spec from being based on AuthConfig v1beta1 to being AuthConfig v1beta2, without anything new (e.g. route selectors)
- Scan through the multiple PRs that added all the other incremental changes to the base branch
- Read some test cases
- Try `make local-setup`ing this branch and running a user guide such as https://github.com/Kuadrant/kuadrant-operator/blob/authpolicy-v2/doc/user-guides/auth-for-app-devs-and-platform-engineers.md
- Try other stuff with AuthPolicies – Here's an inspiration for things you can try: https://github.com/Kuadrant/authorino/blob/main/docs/user-guides.md

### CODE REVIEW

**Comment by Boomatang** [comment]
So far this is looking good. I am not sure if there are verification steps to be carried out right now.

File: bundle/manifests/kuadrant.io_kuadrants.yaml [hunk 1]
```diff
@@ -63,7 +63,6 @@ spec:
                     message:
                       description: message is a human readable message indicating
                         details about the transition. This may be an empty string.
```
Comment (maleck13) [inline bundle/manifests/kuadrant.io_kuadrants.yaml:65]: probably some generated reason, but wondering why this definiton changed?

File: config/crd/bases/kuadrant.io_ratelimitpolicies.yaml [hunk 2]
```diff
@@ -152,8 +155,12 @@ spec:
                                           https://www.rfc-editor.org/rfc/rfc7230.html#section-3.2.2
                                           regarding processing a repeated header,
                                           with special handling for \"Set-Cookie\"."
+                                        maxLength: 256
```
Comment (maleck13) [inline config/crd/bases/kuadrant.io_ratelimitpolicies.yaml:158]: Is this just added as an improvement / fix for RLP?

**Comment by maleck13** [comment]
changes all look reasonable. Bit unsure as to why RLP changed in the PR though ( I am sure there is a good reason)

File: bundle/manifests/kuadrant.io_kuadrants.yaml (hunk 1 above)
Comment (guicassolato) [reply bundle/manifests/kuadrant.io_kuadrants.yaml:65]: Not sure either 🤔

If I run `make generate manifests bundle` in `main`, there are no changes. On this branch on the other hand, it gives me this.

```
❯ go version
go version go1.21.2 darwin/arm64

❯ ./bin/controller-gen --version
Version: v0.10.0

❯ ./bin/operator-sdk version
operator-sdk version: "v1.28.1", commit: "b05f6a56a176a98b7d92c4d4b36076967e0d77f7", kubernetes version: "1.26.0", go version: "go1.19.8", GOOS: "darwin", GOARCH: "arm64"

❯ ./bin/opm version
Version: version.Version{OpmVersion:"v1.26.2", GitCommit:"0e53cafe", BuildDate:"2022-09-21T15:25:25Z", GoOs:"darwin", GoArch:"arm64"}

❯ ./bin/kustomize version
{Version:unknown GitCommit:$Format:%H$ BuildDate:1970-01-01T00:00:00Z GoOs:darwin GoArch:arm64}
```

**Review by maleck13 — APPROVED** [review]

File: doc/rate-limiting.md [hunk 3]
```diff
@@ -2,10 +2,10 @@
 
 A Kuadrant RateLimitPolicy custom resource, often abbreviated "RLP":
 
-1. Allows it to target Gateway API networking resources such as [HTTPRoutes](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRoute) and [Gateways](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.Gateway), using these resources to obtain additional context, i.e., which traffic workload (HTTP attributes, hostnames, user attributes, etc) to rate limit.
-2. Allows to specify which specific subsets of the targeted network resource to apply the limits to.
+1. Targets Gateway API networking resources such as [HTTPRoutes](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRoute) and [Gateways](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.Gateway), using these resources to obtain additional context, i.e., which traffic workload (HTTP attributes, hostnames, user attributes, etc) to rate limit.
+2. Supports targeting subsets (sections) of a network resource resource to apply the limits to.
```
Comment (adam-cattermole) [inline doc/rate-limiting.md:6]: minor nit
```suggestion
2. Supports targeting subsets (sections) of a network resource to apply the limits to.
```

File: doc/reference/authpolicy.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+| `patterns`       | Map<String: [NamedPattern](#namedpattern)>                                                                                                  | No           | Named patterns of lists of `selector`, `operator` and `value` tuples, to be reused in `when` conditions and pattern-matching authorization rules.                                                                                                                              |
+| `when`           | [][PatternExpressionOrRef](https://docs.kuadrant.io/authorino/docs/features/#common-feature-conditions-when)                                | No           | List of additional dynamic conditions (expressions) to activate the policy. Use it for filterring attributes that cannot be expressed in the targeted HTTPRoute's `spec.hostnames` and `spec.rules.matches` fields, or when targeting a Gateway.                               |
```
Comment (adam-cattermole) [inline doc/reference/authpolicy.md:33]:
```suggestion
| `when`           | [][PatternExpressionOrRef](https://docs.kuadrant.io/authorino/docs/features/#common-feature-conditions-when)                                | No           | List of additional dynamic conditions (expressions) to activate the policy. Use it for filtering attributes that cannot be expressed in the targeted HTTPRoute's `spec.hostnames` and `spec.rules.matches` fields, or when targeting a Gateway.                               |
```

Comment (adam-cattermole) [reply doc/reference/authpolicy.md:33]: There's a couple other cases of this typo but definitely not important

**Pull request merged**